src/algorithms/tonal/pitch2midi.cpp

```diff
@@ -0,0 +1,297 @@
+  _timeCompensationBin.resize(0);
+
+  // TODO: this is not clear because it might remove an note_off message which is defined by dnote.
+  //#! it would be better just to provide some code for midiNoteNumbre when this happens
+  if (midiNoteNumber <= 0 && midiNoteNumber >= 127) {
+    //E_INFO("SCAPE");
+    return;
+  }
+
+  // let's define first the message type
+  if (_noteOff) {
+    _messageTypeBin.push_back("note_off");
+  }
+
+  if (_noteOn) {
+    _messageTypeBin.push_back("note_on");
+  }
+
+  if (!_applyCompensation) {
+    onsetTimeCompensation = 0.f;
+    offsetTimeCompensation = 0.f;
+  }
+
+  _midiNoteNumberBin.push_back(static_cast<Real>(dnote_->midiNote));
+  _midiNoteNumberBin.push_back(static_cast<Real>(midiNoteNumber));
+  _timeCompensationBin.push_back(offsetTimeCompensation);
+  _timeCompensationBin.push_back(onsetTimeCompensation);
+    
+  messageTypeOut = _messageTypeBin;
+  midiNoteNumberOut = _midiNoteNumberBin;
+  timeCompensationOut = _timeCompensationBin;
+}
+
+void Pitch2Midi::push(int midiNoteNumber) {
+    // push new MIDI note number in the MIDI buffer
+    _midiNoteNumberVector[0] = static_cast<Real>(midiNoteNumber);
+    _framebuffer->input("frame").set(_midiNoteNumberVector);
+    _framebuffer->output("frame").set(_buffer);
+    _framebuffer->compute();
+}
+
+void Pitch2Midi::compute()
+{
+  // former MidiPool inputs are now Pitch2Midi internal vars
+  // all we need is to run the conversions:
+  const Real& pitch = _pitch.get();
+  const int& voiced = _voiced.get();
+    
+  // do sanity checks
+  if (pitch < 0) {
+    throw EssentiaException("PitchContoursMultiMelody: specified duration of the input signal must be non-negative");
+  }
+
+  getMidiNoteNumber(pitch);
+    
+  // refresh note_on and note_off timestamps
+  _noteOn = false;
+  _noteOff = false;
+  
+  // unvoiced frame detection
+  if (!voiced) {
+    if ( _NOTED_ON ) {
+      _unvoicedFrameCounter++;
+      if (_unvoicedFrameCounter > _minNoteChangeThreshold) {
+        _NOTED_ON = false;
+        _noteOff = true;
+        updateDnote();
+        setOutputs(dnote_->midiNote, 0.f, _minNoteChangePeriod);
+        //E_INFO("offset(unvoiced frame)");
+        _unvoicedFrameCounter = 0;
+        _offsetCheckCounter = 0;
+        _onsetCheckCounter = 0;
+      }
+    } else {
+      _unvoicedFrameCounter = 0;
+      push(0);    // push 0th MIDI note to remove the past
+      _offsetCheckCounter = 0;
+      _onsetCheckCounter = 0;
+    }
+    return;
+  }
+
+  _unvoicedFrameCounter = 0;
+    
+  // push new MIDI note number in the MIDI buffer
+  push(_midiNoteNumberTransposed);
+
+  // update max_voting
+  getMaxVoted();
+
+  /*
+  E_INFO("onset thresholds: "<< _onsetCheckCounter <<" - " << _minOnsetCheckThreshold);
+  E_INFO("offset thresholds: "<< _offsetCheckCounter <<" - " << _minOffsetCheckThreshold);
+  */
+
+  // analyze pitch buffer
+  if (hasCoherence() && _NOTED_ON) {
+    if (isCurrentMidiNoteEqualToMaxVoted()) {
```

Comment:
Do we need this function as well? Takes more time to check what it does and more boilerplate code to define it.

Reply:
Which one? `isCurrentMidiNoteEqualToMaxVoted()`? I would remove it, this is a Jorge's definition. For me it is easier to read in the code.

Reply:
Yes, I am referring to this function.

Reply:
I would do the same for `isMaxVotedZero()` and `isMaxVotedCountGreaterThanMinOcurrenceRate()` what do you think?